velox/common/base/SimdUtil-inl.h

```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
   return true;
 }
 
+namespace detail {
+template <typename T>
+T clearLeftmostSet(const T value) {
+  assert(value != 0);
+
+  return value & (value - 1);
+}
+
+template <typename T>
+unsigned FOLLY_ALWAYS_INLINE getFirstBitSet(const T value) {
+  assert(value != 0);
+
+  return __builtin_ctz(value);
+}
+
+template <>
+unsigned FOLLY_ALWAYS_INLINE getFirstBitSet<uint64_t>(const uint64_t value) {
+  assert(value != 0);
+
+  return __builtin_ctzl(value);
+}
+
+#if XSIMD_WITH_AVX2
+// AVX2 is faster than sse2
+#define SIMD_STRSTR
+using CharVector = xsimd::batch<uint8_t, xsimd::avx2>;
+#elif XSIMD_WITH_NEON
+#define SIMD_STRSTR
+using CharVector = xsimd::batch<uint8_t, xsimd::neon>;
+#endif
+
+#ifdef SIMD_STRSTR
+size_t FOLLY_ALWAYS_INLINE
+smidStrstrAnysize(const char* s, size_t n, const char* needle, size_t k) {
+  const auto first = CharVector::broadcast(needle[0]);
+  const auto last = CharVector::broadcast(needle[k - 1]);
+
+  for (size_t i = 0; i < n; i += CharVector::size) {
+    const auto block_first = CharVector::load_unaligned(s + i);
+    const auto block_last = CharVector::load_unaligned(s + i + k - 1);
+
+    const auto eq_first = (first == block_first);
+    const auto eq_last = (last == block_last);
+
+    auto mask = toBitMask(eq_first && eq_last);
+    ;
+
+    while (mask != 0) {
+      const auto bitpos = detail::getFirstBitSet(mask);
+
+      if (memcmp(s + i + bitpos + 1, needle + 1, k - 2) == 0) {
+        return i + bitpos;
+      }
+
+      mask = detail::clearLeftmostSet(mask);
+    }
+  }
+
+  return std::string::npos;
+}
+
+template <size_t k, typename MEMCMP>
+size_t FOLLY_ALWAYS_INLINE smidStrstrMemcmp(
+    const char* s,
+    size_t n,
+    const char* needle,
+    MEMCMP memcmp_fun) {
+  assert(k > 0);
+  assert(n > 0);
+
+  auto first = CharVector::broadcast(needle[0]);
+  auto last = CharVector::broadcast(needle[k - 1]);
+  for (size_t i = 0; i < n; i += CharVector::size) {
```

Comment:
`i + CharVector::size <= n` otherwise it's overreading, you need to handle trailing characters separately